File: implementations/react/src/patterns/behavioral/visitor/element-grocery.ts
import type { Visitor } from "./visitor";
import type { Element } from "./element";

/**
 * Concrete Element que representa un artículo de supermercado
 */

export class Grocery implements Element {
  public name: string;
  public unitPrice: number;
  public qty: number;
  public perishable: boolean;

  constructor(
    name: string,
    unitPrice: number,
    qty: number,
    perishable = true
  ) {
    this.name = name;
    this.unitPrice = unitPrice;
    this.qty = qty;
    this.perishable = perishable;
  }

  accept(v: Visitor) { v.visitGrocery(this); }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/visitor/element_grocery.dart
import 'element.dart';
import 'visitor.dart';

class Grocery implements Element {
  final String name;
  final double unitPrice;
  int qty;
  final bool perishable;

  Grocery(this.name, this.unitPrice, this.qty, [this.perishable = true]);

  @override
  void accept(Visitor v) => v.visitGrocery(this);
}
